nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-kubernetes-bundle/nifi-framework-kubernetes-leader-election/src/main/java/org/apache/nifi/kubernetes/leader/election/KubernetesLeaderElectionManager.java

```diff
@@ -0,0 +1,316 @@
+        return isParticipating(participantId);
+    }
+
+    /**
+     * Get Leader Identifier for Role
+     *
+     * @param roleName Role Name for requested Leader Identifier
+     * @return Leader Identifier or null when not found
+     */
+    @Override
+    public String getLeader(final String roleName) {
+        requireRoleName(roleName);
+
+        final long pollStarted = System.nanoTime();
+        try {
+            final String leader;
+            if (started.get()) {
+                final String roleLeader = roleLeaders.get(roleName);
+                if (roleLeader == null) {
+                    logger.debug("Leader not registered: finding Leader for Role [{}]", roleName);
+                    leader = leaderElectionCommandProvider.findLeader(roleName);
+                } else {
+                    leader = roleLeader;
+                }
+            } else {
+                logger.debug("Manager not running: finding Leader for Role [{}]", roleName);
+                leader = leaderElectionCommandProvider.findLeader(roleName);
+            }
+            return leader;
+        } finally {
+            final long elapsed = System.nanoTime() - pollStarted;
+            registerPollTime(elapsed);
+        }
+    }
+
+    /**
+     * Determine whether current node is the Leader for the specified Role
+     *
+     * @param roleName Role Name to be evaluated
+     * @return Leader Status
+     */
+    @Override
+    public boolean isLeader(final String roleName) {
+        requireRoleName(roleName);
+        final boolean leader;
+
+        final String participantId = getParticipantId(roleName);
+        if (participantId == null) {
+            logger.debug("Role [{}] not participating in Leader election", roleName);
+            leader = false;
+        } else {
+            final String leaderId = getLeader(roleName);
+            leader = participantId.equals(leaderId);
+            if (leader) {
+                logger.debug("Role [{}] Participant ID [{}] is Leader", roleName, participantId);
+            } else {
+                logger.debug("Role [{}] Participant ID [{}] not Leader", roleName, leaderId);
+            }
+        }
+        return leader;
+    }
+
+    protected ExecutorService createExecutorService() {
+        return Executors.newFixedThreadPool(SERVICE_THREADS, new NamedThreadFactory());
+    }
+
+    protected LeaderElectionCommandProvider createLeaderElectionCommandProvider() {
+        final NamespaceProvider namespaceProvider = new ServiceAccountNamespaceProvider();
+        final String namespace = namespaceProvider.getNamespace();
+        final KubernetesClientProvider kubernetesClientProvider = new StandardKubernetesClientProvider();
+        return new StandardLeaderElectionCommandProvider(kubernetesClientProvider, namespace);
+    }
+
+    private void registerLeaderElectionCommand(final String roleName, final LeaderElectionStateChangeListener listener, final String participantId) {
```

Comment:
There's a lot of check-then-modify here. I think we need to ensure that this is thread safe and synchronize the method or add locking

Reply:
This method is only called from `register`, which is marked as `synchronized`, do you think any additional synchronization is necessary?

Reply:
Ah ok, good call. So technically it's correct/safe. But I would generally mark this as `synchronized` as well because it has virtually zero cost, since the synchronization lock has already been obtained, but it makes it more clear that the actions can only be taken while synchronized. Or alternatively documenting via JavaDoc that it should only be called while synchronized. But I'd prefer just synchronizing the method itself.